Core/HW/SasAudio.cpp

```diff
@@ -493,47 +483,28 @@ void SasInstance::MixVoice(SasVoice &voice) {
 			break;
 		// else fallthrough! Don't change the check above.
 	default:
-		// Load resample history (so we can use a wide filter)
-		resampleBuffer[0] = voice.resampleHist[0];
-		resampleBuffer[1] = voice.resampleHist[1];
-
-		// Figure out number of samples to read.
-		// Actually this is not entirely correct - we need to get one extra sample, and store it
-		// for the next time around. A little complicated...
-		// But for now, see Smoothness HACKERY below :P
-		u32 numSamples = ((u32)voice.sampleFrac + (u32)grainSize * (u32)voice.pitch) >> PSP_SAS_PITCH_BASE_SHIFT;
-		if ((int)numSamples > grainSize * 4) {
-			ERROR_LOG(SASMIX, "numSamples too large, clamping: %i vs %i", numSamples, grainSize * 4);
-			numSamples = grainSize * 4;
-		}
-
 		// This feels a bit hacky.  The first 32 samples after a keyon are 0s.
-		const bool ignorePitch = voice.type == VOICETYPE_PCM && voice.pitch > PSP_SAS_PITCH_BASE;
+		int delay = 0;
 		if (voice.envelope.NeedsKeyOn()) {
-			int delay = ignorePitch ? 32 : (32 * (u32)voice.pitch) >> PSP_SAS_PITCH_BASE_SHIFT;
+			const bool ignorePitch = voice.type == VOICETYPE_PCM && voice.pitch > PSP_SAS_PITCH_BASE;
+			delay = ignorePitch ? 32 : (32 * (u32)voice.pitch) >> PSP_SAS_PITCH_BASE_SHIFT;
 			// VAG seems to have an extra sample delay (not shared by PCM.)
 			if (voice.type == VOICETYPE_VAG)
 				++delay;
-			voice.ReadSamples(resampleBuffer + 2 + delay, numSamples - delay);
-		} else {
-			voice.ReadSamples(resampleBuffer + 2, numSamples);
 		}
 
-		// Smoothness HACKERY
-		resampleBuffer[2 + numSamples] = resampleBuffer[2 + numSamples - 1];
-
-		// Save resample history
-		voice.resampleHist[0] = resampleBuffer[2 + numSamples - 2];
-		voice.resampleHist[1] = resampleBuffer[2 + numSamples - 1];
-
 		// Resample to the correct pitch, writing exactly "grainSize" samples.
-		// This is a HORRIBLE resampler by the way.
 		// TODO: Special case no-resample case (and 2x and 0.5x) for speed, it's not uncommon
-
 		u32 sampleFrac = voice.sampleFrac;
-		for (int i = 0; i < grainSize; i++) {
-			// For now: nearest neighbour, not even using the resample history at all.
-			int sample = resampleBuffer[sampleFrac / PSP_SAS_PITCH_BASE + 2];
+		for (int i = delay; i < grainSize; i++) {
+			while (sampleFrac >= PSP_SAS_PITCH_BASE) {
+				voice.resampleHist[0] = voice.resampleHist[1];
+				voice.ReadSamples(&voice.resampleHist[1], 1);
```

Comment:
It could, which is what "Not sure about the performance impact, but calling ReadSamples with 1 as an argument can't be terribly fast" in the pull request description refers to :)